[ADD] 8.0 purchase request modules

JordiBForgeFlow opened the pull request:
This PR adds a new entity 'Purchase Request'.

The following modules are added:

purchase_request
purchase_request_to_requisition
purchase_request_to_rfq
# Purchase Request

You use this module if you wish to give notification of requirements of
materials and/or external services and keep track of such requirements.

Requests can be created either directly or indirectly. "Directly" means that
someone from the requesting department enters a purchase request manually.

The person creating the requisition determines what and how much to order,
and the requested date.

"Indirectly" means that the purchase request initiated by the application
automatically, for example, from procurement orders.

A purchase request is an instruction to Purchasing to procure a certain
quantity of materials services, so that they are is available at a
certain point in time.

A line of a requisition contains the quantity and requested date of the
material to be supplied or the quantity of the service to be performed. You
can indicate the service specifications if needed.

It is possible to print the purchase request.
# Purchase Request to Calls for Bid

This module introduces the following features:

The possibility to create new Calls for Bids or update existing Bids from Purchase Request Lines.
The possibility to create Purchase Requests on the basis of Procurement Orders.
# Purchase Request to RFQ

This module adds the possibility to create or update Requests for
Quotation (RFQ) from Purchase Request Lines.

Commit:
[ADD] Module to split purchase receptions by dates

Commit:
[MIGR] purchase_delivery_split_date

Use v8 API
Follow the OCA guidelines for manifest and README
PEP8
Extend _create_stock_move because the method we used to extend in v7 doesn't exist anymore.

Commit:
[IMP] Unit test for purchase_delivery_split_date

eLBati committed:
[ADD] test that  cancel an existing demo PO and create a new revision

eLBati committed:
[ADD] fill_unrevisioned_name post init hook

JordiBForgeFlow committed:
[ADD] purchase request modules

JordiBForgeFlow committed:
[ADD] purchase request modules

JordiBForgeFlow commented:
This is the port of https://github.com/OCA/purchase-workflow/pull/186 to 8.0

JordiBForgeFlow committed:
[IMP] Add try me on runbot

JordiBForgeFlow committed:
[IMP] Add try me on runbot

Commit:
[FIX] Use suspend_security on search on purchase order for users that have no access right on purchases.

Commit:
[IMP][purchase_partial_invoicing] Add dependencies informations in readme file

sbidoul committed:
[ADD] addons table in README.md

JordiBForgeFlow committed:
[FIX] Various fixes

JordiBForgeFlow committed:
[FIX] Various fixes

robertmooreg commented:
Good morning, 
I have installed the PR and discovered that when a sale order is placed, the workflow doesn´t generate a withdrawal note,may this be an issue or just a configuration problem?

Thank you in advanced.

JordiBForgeFlow commented:
Did you test on the latest commit?
http://3115873-189-87e07c.runbot1.odoo-community.org/web?db=3115873-189-87e07c-all#id=1&view_type=form&model=purchase.request&menu_id=329&action=398

I just reviewed the flow and it posts the message.

![image](https://cloud.githubusercontent.com/assets/7683926/10099803/69b67f10-638d-11e5-899c-53dbe1bcf766.png)

robertmooreg commented:
The problem is that when a sale order is placed, the workflow doesn´t generate a delivery order

what might causing this?

JordiBForgeFlow commented:
A sales order? what do you mean? Purchase Requests are for purchasing, not
for sales. Please clarify

On Fri, Sep 25, 2015 at 2:04 PM, robertmooreg notifications@github.com
wrote:

> The proble is that when a sale order is placed, the workflow doesn´t
> generate a withdrawal
> 
> what might causing this?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/purchase-workflow/pull/189#issuecomment-143197609
> .

## 

Jordi Ballester Alomar
Founder | Eficent
(+34) 629530707 | jordi.ballester@eficent.com | http://www.eficent.com
Twitter: https://twitter.com/jbeficent_erp | Linkedin:
https://www.linkedin.com/in/jordiballesteralomar

liebana commented:
Sorry for the misunderstanding Jordi. What Robert means is that somehow the sale workflow is broken when your modules are installed. We've tested in the runbot provided and in a local installation, and in both enviroments the delivery order is not automatically generated when confirming the sale order.

We can't give you further technical details as my development team has not been able to study this deeply, but you can test yourself with the creation of any sale order (with any workflow configuration regarding invoicing or delivery) in runbot.

JordiBForgeFlow commented:
I just tested on runbot and saw that the delivery order was not created at all. Then I uninstalled the modules and installed them again, and now works:
http://3115873-189-87e07c.runbot1.odoo-community.org

Perhaps the issue is related to how is the runbot building the database, because the second time I installed the application started the accounting wizard. 

Creating a fresh database in my local server this issue did not reproduce.

JordiBForgeFlow committed:
[FIX] fixed travis errors

JordiBForgeFlow committed:
[FIX] fixed travis errors

liebana commented:
It seems that the module purchase_request_to_requisition is the troublesome one.

hugosantosred commented on purchase_request_to_requisition/models/procurement.py:
```diff
@@ -0,0 +1,69 @@
+            req = request_obj.create(request_data)
+            procurement.message_post(body=_("Purchase Request created"))
+            procurement.request_id = req.id
+        return True
```
I think a call to super in _run is missing to execute the other conditions defined in other addons that uses this method too.
This is why is not  creating the pickings for Sale Orders with purchase_request_to_requisition installed.

miketelahun committed:
FIX: properly propagate _run() method of procurement.order accross the class hierarchy

miketelahun committed:
FIX: properly propagate _run() method of procurement.order accross the class hierarchy

mtelahun commented:
The module purchase_request_to_requisition overrides the _run() method of procurement.order but doesn't propagate it through to the rest of the class hierarchy when it's finished. I opened a pull request to fix this: https://github.com/Eficent/purchase-workflow/pull/1

Tested and verified locally.

JordiBForgeFlow committed:
Merge pull request #1 from mtelahun/fix-eficent-pr-189-procurement-order-run-super

FIX: properly propagate _run() method of procurement.order across the class hierarchy

JordiBForgeFlow commented:
Hi, @miketelahun can you sign the CLA as requested by the clabot?

JordiBForgeFlow commented:
At the time of install the module should make 'False' the values of 'purchase_requisition' field in the product_template, because procurement orders will now create purchase requests instead of requisitions. I'm not sure how can this be achieved. Can anyone help me on this?

JordiBForgeFlow committed:
[FIX] travis errors

JordiBForgeFlow committed:
[FIX] travis errors

JordiBForgeFlow committed:
[FIX] openerp.models: ir.values.create() with unknown fields: object

JordiBForgeFlow committed:
[FIX] openerp.models: ir.values.create() with unknown fields: object

mtelahun commented:
Ok. Will do.

mtelahun commented:
@jbeficent I assume you mean that any products that have purchase_requisition field set on the procurement tab need to  have it removed.

If installing the purchase request modules means that we no longer go directly to purchase requisition, then the easiest solution would be to re-use purchase_requisition field instead of creating a new purchase_request field.  Then in the view we just change string for purchase_requisition from 'Call for Bids' to 'Purchase Request.' However, if purchase_requisition and purchase_request must co-exist in the same database then maybe a wizard allowing the user to change the products manually would be the ideal solution.

sysadminmatmoz commented:
Do not forget about this PR; it's a great functionality that I already use on my dev databases and planning to launch on several production databases this night.

rafaelbn commented:
@jbeficent Why you create a new menu? This is an improvement for purchases and you have already add this to purchases. Why to have it duplicated?

JordiBForgeFlow commented:
Any employee can initiate the purchase of goods, nor only the purchase user or manager. Putting the menu into Purchasing would require to modify the standard Odoo Purchasing menu, so that everybody can access to it.

I'd appreciate your views on this. Perhaps you're right. I took as an example the module by Vauxoo purchase_requisition_everyone that does the same. Allows any employee to create requisitions, and adds an extra mnu

JordiBForgeFlow committed:
[IMP] Added message notifications on change of the status of the purchase request.

JordiBForgeFlow committed:
[IMP] Added message notifications on change of the status of the purchase request.

rafaelbn commented:
In my opinion the correct way would be create a new ACL. In this module you can add that permission (only the ones needed) to employees. You can check here how is done for projects https://github.com/OCA/project/tree/8.0/project_baseuser.
You don't need yo create a separate module, just add this ACL to employees for this access.

JordiBForgeFlow commented:
@rafaelbn But how will they access to the menu entries? What is the root menu they should use to access to Purchase Requests? ACL's are helpful to define who has permissions to read, write, delete an object, but will not by themselves offer new menu entires.

Take into account that the root menu "Purchases" is restricted to groups "Purchase User" and "Purchase Manager":

```
    <menuitem name="Purchases"
        id="base.menu_purchase_root"
        groups="group_purchase_manager,group_purchase_user"
        sequence="60"/>
```

JordiBForgeFlow committed:
[IMP] purchase_request_to_requisition. We don't allow to set call for bids and purchase request at the same time in the product form.

JordiBForgeFlow committed:
[IMP] purchase_request_to_requisition. We don't allow to set call for bids and purchase request at the same time in the product form.

rafaelbn commented:
Could you create a new group: group_purchase_employee with permisión only to purchase request and and this group to that menu item "Purchases"?

JordiBForgeFlow commented:
That can be a good idea. But then we need a module "purchase_baseuser", similar to "project_baseuser" that is used to manage the ACL and menu accesses for Purchasing. 
That way, everybody that needs to add a new menu entry within Purchasing, can include the dependency to this module.

rafaelbn commented:
Maybe the opinion of an expert could help us... cc @dreispt @pedrobaeza @StefanRijnhart ?

sbidoul committed:
[UPD] prefix versions with 8.0

JordiBForgeFlow committed:
[FIX] fixed group by error

JordiBForgeFlow committed:
[FIX] fixed group by error

JordiBForgeFlow committed:
[IMP] extend purchase request line search

JordiBForgeFlow committed:
[IMP] extend purchase request line search

JordiBForgeFlow committed:
[FIX] fixed search filter

JordiBForgeFlow committed:
[FIX] fixed search filter

sbidoul committed:
[MOV] move addons out of __unported__ (they remain not installable)

okuryan commented:
Hi @jbeficent , @rafaelbn. Wanted to check if that PR still alive. Odoo really lacks this functionality so will be good to have it in OCA

JordiBForgeFlow commented:
@okuryan, yes it is very well alive. Would be good to have your feedback and thumbs up if ok.

On my side I identified the need to revisit the link of procurement order to purchase request, so that the procurement orders are a many2one in purchase request line. When a PO is created from a Purchase Request the associated procurements should move to the PO, so that the procurement is kept up to date with the associated PO line.

yvaucher committed:
Merge pull request #183 from eLBati/8.0-fix_purchase_order_revision-test

fix purchase order revision test and post_init_hook

sbidoul commented on purchase_request_to_requisition/models/procurement.py:
```diff
@@ -0,0 +1,70 @@
+        return {
+            'origin': procurement.origin,
+            'company_id': procurement.company_id.id,
+            'warehouse_id': warehouse_id,
```
must be `warehouse_id.id`

sbidoul commented on purchase_request_to_requisition/models/procurement.py:
```diff
@@ -0,0 +1,70 @@
+            'origin': procurement.origin,
+            'company_id': procurement.company_id.id,
+            'warehouse_id': warehouse_id,
+            'line_ids': [(4, request_line_data)]
```
must be `'line_ids': [(0, None, request_line_data)]`

sbidoul commented on purchase_request_to_requisition/README.rst:
```diff
@@ -0,0 +1,86 @@
+------------------------------------------------
+Go to the product form, in tab 'Procurement' and choose 'Purchase Request'.
+When a procurement order is created and the supply method is 'Make to Order'
+the application will now create a Purchase Request.
```
Question: why is this capability to generate purchase requests from purchase order in this purchase_request_to_requisition module? This capability is useful even if we don't use purchase_requisition. Would it not be better in a stand alone module?

sbidoul commented:
Hi @jbeficent 

I did a few tests of this module, and it looks great! Thanks for this contribution. We may need it next year.

What you explain in https://github.com/OCA/purchase-workflow/pull/189#issuecomment-150211829 looks promising too.

I posted a question and 2 fixes suggestion (procurement not working without these).

Last question: is it on purpose that you show quantities of other user's requests in "Quantity in RFQ or PO".

JordiBForgeFlow commented:
On Wed, Oct 28, 2015 at 3:12 PM, Stéphane Bidoul (ACSONE) <
notifications@github.com> wrote:

> Hi @jbeficent https://github.com/jbeficent
> 
> I did a few tests of this module, and it looks great! Thanks for this
> contribution. We may need it next year.
> 
> What you explain in #189 (comment)
> https://github.com/OCA/purchase-workflow/pull/189#issuecomment-150211829
> looks promising too.
> 
> I posted a question and 2 fixes suggestion (procurement not working
> without these).
> 
> Thanks, I will review them.
> 
> Last question: is it on purpose that you show quantities of other user's
> requests in "Quantity in RFQ or PO".
> 
> This looks like a bug. Will also review.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/purchase-workflow/pull/189#issuecomment-151857452
> .

## 

Jordi Ballester Alomar
Founder | Eficent
(+34) 629530707 | jordi.ballester@eficent.com | http://www.eficent.com
Twitter: https://twitter.com/jbeficent_erp | Linkedin:
https://www.linkedin.com/in/jordiballesteralomar

Commit:
[FIX] manifest conventions

JordiBForgeFlow commented:
Please note, every friday at 11AM CET we're holding a webinar to explain this module. Free to OCA members. Join here: https://www.eventbrite.com/e/streamline-your-purchasing-process-in-odoo-using-purchase-requests-europe-time-zone-tickets-18754471128

Commit:
[TEST][purchase_partial_invoicing] Call ir.rule register hook

Commit:
[FIX] do not display product_tmpl_id on product_supplier_info form if opened from product form

Commit:
[FIX][product_supplierinfo_discount] Fail when same products has multiple records of same supplier

Odoo does not restrict entering mulltiple product.supplierinfo records
for the same product and same partner.

Commit:
[FIX][product_supplierinfo_discount] Less code to select first element

pedrobaeza committed:
Merge pull request #195 from acsone/product_supplier_fix_cpi

[8.0] Product By Supplier issue

yvaucher committed:
Merge branch 'acsone-8.0-fix-access-parial-invoicing-ape' into 8.0

OCA-git-bot committed:
[UPD] addons table in README.md

pedrobaeza committed:
Merge pull request #196 from andrius-preimantas/fix_supplierinfo_discount_singleton

[FIX][product_supplierinfo_discount] Fail when same products has mult…

JordiBForgeFlow committed:
[IMP] split from purchase_request_to_requisition the module purchase_request_procurement

JordiBForgeFlow committed:
[IMP] split from purchase_request_to_requisition the module purchase_request_procurement

JordiBForgeFlow committed:
[FIX] pep8 issues

JordiBForgeFlow committed:
[FIX] pep8 issues

eLBati committed:
FIX

1.create purchase order ..maybe: PO00001
2.confirm purchase order
3. cancle this purchase order .name: PO00001
4.new revision ....then create new purhcase order name i s PO00001-1
5.cancel this purchase order name PO00001-1
6.more->duplicate
7.new purchase order create...name is PO00002
8. confirm this order PO00002
9.cancel this order PO00002
10. new revision.....then error

Order Reference must be unique per Company

kittiu commented:
@jbeficent 

+1,  finally we have a serious Purchase Request :)

In v7, there is a (joke) Purchase Requisition menu, no matter how we look at it, it is not a Purchase Requisition in the real business sense. Thank God, v8 change name to Call for Bid but still do the exact same thing. Which means there is no Purchase Requisition any more in v8.

The real Purchase Requisition should give "The Power of Negotiation"  into purchaser hand. The flow should be flexible as everyone has explained (splitting lines, grouping lines form many PRs, tracibility, etc). For smaller organization, may not required, but for large organization, it is seriously needed.

**Why can't we extend the existing Call for Bid (purchase_requisition) to cover more cases?**

We did! (https://github.com/kittiu/addons-eco/tree/7.0/purchase_requisition_multi_supplier) but it make the code very complex and still can't cover all cases. Such as we can't select lines form multiple requests.

When I test this Pull Request, it does cover more use cases and at the same time the code is much simpler, which is the result of having a new model.

**So I see the separate model a necessity as well**

JordiBForgeFlow commented:
@sbidoul I attended your comments. I think that this one is ready to review & merge.

Commit:
[FIX] fix error when create PO/Bid from PR line w/o product_id

Commit:
[FIX] fix error when create PO/Bid from PR line w/o product_id

JordiBForgeFlow committed:
Merge pull request #2 from kittiu/8.0-purchase_request

[FIX] fix error when create PO/Bid from PR line w/o product_id

pedrobaeza committed:
Merge pull request #210 from eLBati/fix_unrevisioned_name

FIX purchase_order_revision

JordiBForgeFlow committed:
[FIX] adjustments to make the purchase request compatible with v8

JordiBForgeFlow committed:
[FIX] adjustments to make the purchase request compatible with v8

JordiBForgeFlow committed:
[FIX] merged changes

Commit:
[FIX] fix related to picking type and warehouse for v8

Minimum fix just to continue on purchase_requistion_operating_unit
Otherwise, it just give me errors.
Please review and make fix to your branch and I will follow.

Commit:
[FIX] fix related to picking type and warehouse for v8

Minimum fix just to continue on purchase_requistion_operating_unit
Otherwise, it just give me errors.
Please review and make fix to your branch and I will follow.

JordiBForgeFlow committed:
Merge pull request #3 from kittiu/8.0-purchase_request

[FIX] fix related to picking type and warehouse for v8

Commit:
[FIX] recover what has fixed from previous PR (was removed)

https://github.com/Eficent/purchase-workflow/pull/2

Commit:
[FIX] recover what has fixed from previous PR (was removed)

https://github.com/Eficent/purchase-workflow/pull/2

Commit:
Merge branch '8.0-purchase_request' of git@github.com:Eficent/purchase-workflow.git into 8.0-purchase_request

JordiBForgeFlow committed:
Merge pull request #4 from kittiu/8.0-purchase_request

8.0 purchase request

Commit:
[FIX] purchase_request_to_rfq, assign date_planned if none.

If no product_id, date_planned won't be prepare. But we need it or
error.

Commit:
[ENH] purchase_request, PR Line only filtered with non-Draft state

Commit:
[ENH] purchase_request, PR Line only filtered with non-Draft state

pedrobaeza committed:
[IMP] .travis.yml: Coveralls+Transifex activation

chafique-delli committed:
add purchase_picking_state module

chafique-delli committed:
add french translate

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[IMP] purchase_discount: load price with discount in stock_move when confirm a purchase order

elicoidal committed:
Merge pull request #217 from agaldona/8.0-purchase_discount

[IMP] purchase_discount: load price with discount in stock_move when confirm a purchase order

chafique-delli committed:
fix README.rst and fix openerp.py and refactor _compute_picking_state method and add tests and add invisible attribut for picking_state field in views

chafique-delli committed:
add icon.png

chafique-delli committed:
fix tests

Commit:
Conform to OCA standards

pedrobaeza committed:
Merge pull request #218 from akretion/8-purchase-picking-state

8 purchase picking state

OCA-git-bot committed:
[UPD] addons table in README.md

oca-transbot committed:
OCA Transbot updated translations from Transifex

gurneyalex committed:
Merge pull request #153 from numerigraphe/8.0-add-purchase_delivery_split_date

8.0 add purchase delivery split date

OCA-git-bot committed:
[UPD] addons table in README.md

mikevhe18 committed:
[ADD]purchase_rfq_number

mikevhe18 committed:
[UPD]Changing url for issue on README.rst

mikevhe18 committed:
[FIX]Fix errors in writing

mikevhe18 committed:
[UPD]Creating Unittest

mikevhe18 committed:
Change Block Comment from '#' into '# '

OCA-git-bot committed:
[ADD] setup.py

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[FIX] a line with a bigger invoiced quantity than the quantity ordered should be considered as invoiced.

Commit:
[FIX] Error raised if no procurement during the creation of the purchase order form the purchase request

Commit:
[FIX] Error raised if no procurement during the creation of the purchase order form the purchase request

JordiBForgeFlow committed:
Merge pull request #7 from acsone/8.0-purchase_request-jne

[FIX] Error raised if no procurement during the creation of the purch…

yvaucher committed:
Merge pull request #229 from acsone/8.0-fix-purchase-partial-quantity-ape

[FIX] a line with a bigger invoiced quantity than the quantity ordered should be considered as invoiced.

yvaucher committed:
Merge pull request #223 from mikevhe18/8.0-purchase_rfq_number

[8.0]purchase_rfq_number

OCA-git-bot committed:
[UPD] addons table in README.md

OCA-git-bot committed:
[ADD] setup.py

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[ADD] translation files .pot

Commit:
[ADD] translation files .pot

JonathanNEMRY commented:
:+1:  functional tests

JordiBForgeFlow committed:
Merge pull request #8 from acsone/8.0-purchase_request-fr_translations-jne

[ADD] translation files .pot

oca-transbot committed:
OCA Transbot updated translations from Transifex

JonathanNEMRY commented:
@jbeficent I'll try to make a technical review too (even if I already used the main part of the given source code without any problem). Pending this what are the missing points into this PR that are required to merge it?

JordiBForgeFlow committed:
[FIX] messages to the purchase request generated in purchase order, stock picking and purchase requisition should notify to followers of the purchase request.

JordiBForgeFlow committed:
[FIX] messages to the purchase request generated in purchase order, stock picking and purchase requisition should notify to followers of the purchase request.

JordiBForgeFlow committed:
Merge remote-tracking branch 'origin/8.0-purchase_request' into 8.0-purchase_request

JordiBForgeFlow commented:
@JonathanNEMRY I just added a new commit related to sending notifications to followers of the purchase request when the PO is confirmed and when the incoming shipment is completed.

Aside from that, the peding tasks are:
- Greenify Travis
- Possibly review YML tests (Perhaps they need to be adjusted, after the latest changes were made to the module).

mikevhe18 committed:
[ADD] purchase_requisition_type (#231)

* [ADD] purchase_requisition_type

* Delete duplicate usage on README.rst

* Fix error when encode BID

* [UPD] Add invoice_method on _prepare_purchase_order function

* [UPD] Add invoice_method on unittest prepare_purchase_order

* Fixing flake error E501

OCA-git-bot committed:
[UPD] addons table in README.md

OCA-git-bot committed:
[ADD] setup.py

JonathanNEMRY commented:
Ok I propose myself to support those two last points .
About a 9.0 version is there already something done?
If not and as this feature seems very liked by customers we can propose to port it to v9 during the next few months.
I've created an issue about the migration from 8.0 to 9.0 here https://github.com/OCA/purchase-workflow/issues/232

Please let me know your opinion/suggestions about this topic 

Jonathan

JordiBForgeFlow commented:
@JonathanNEMRY We have it working locally in v9, but it needs incorporating the latest changes made in v8. In can be up to date quickly. I was planning to release the v9 as LGPL. Would other contributors agree?

sbidoul commented:
@jbeficent ACSONE's contributions are small and you did most of this great work. So no issue from our side if you change it to LGPL.

From our side, functional tests are good and :+1:  to merge 8.0 when CI is green.

Commit:
[CHG] greenify travis
* Move yml into python test
* Pep8/Flake8

Commit:
[CHG] greenify travis
* Move yml into python test
* Pep8/Flake8

Commit:
[CHG] greenify travis and port ymal tests to python

Commit:
[CHG] greenify travis and port ymal tests to python

oca-transbot committed:
OCA Transbot updated translations from Transifex

JonathanNEMRY commented:
@jbeficent what do you think about my last comments? don't hesitate If anyone else has an other idea
https://github.com/Eficent/purchase-workflow/pull/9

Commit:
[CHG] add dependency on purchase_request_to_rfq
see also https://github.com/Eficent/purchase-workflow/pull/9#issuecomment-222669079

Commit:
[CHG] add dependency on purchase_request_to_rfq
see also https://github.com/Eficent/purchase-workflow/pull/9#issuecomment-222669079

JordiBForgeFlow committed:
Merge pull request #9 from acsone/8.0-greenify_travis_purchase_request-jne

8.0 greenify travis purchase request jne

JordiBForgeFlow committed:
Merge pull request #5 from kittiu/8.0-purchase_request

[FIX] purchase_request_to_rfq, assign date_planned if none.

Commit:
[FIX] travis

Commit:
[FIX] travis

Commit:
[CHG] restart build

Commit:
[CHG] restart build

JordiBForgeFlow committed:
Merge pull request #10 from acsone/8.0-greenify_travis_purchase_request-jne

[FIX] travis

JonathanNEMRY commented:
Test are red because of https://travis-ci.org/OCA/purchase-workflow/jobs/132452375 merged into https://github.com/OCA/purchase-workflow/pull/231
cc @moylop260 @mikevhe18 please could you have a look on this?

@jbeficent thanks for your merges we 're green now for purchase request modules

mikevhe18 commented:
Hi, @JonathanNEMRY i've fix it #235 

Thanks in advance

mikevhe18 committed:
[REV]test_prepare_purchase_order.py (#235)

moylop260 commented:
The PR https://github.com/OCA/purchase-workflow/pull/235 is merged.
Could you rebase this one?

moylop260 commented:
FYI odoo recommended us don't use `api.one` https://github.com/odoo/odoo/pull/8527
You can see the beta result of pylint-odoo [here](https://travis-ci.org/OCA/purchase-workflow/jobs/134377023#L418-L439)

``` bash
************* Module purchase_request_procurement.models.procurement
purchase_request_procurement/models/procurement.py:16: [W8104(api-one-deprecated), Procurement.copy] api.one deprecated
************* Module purchase_request_to_rfq.models.purchase_order
purchase_request_to_rfq/models/purchase_order.py:98: [W8104(api-one-deprecated), PurchaseOrderLine._has_purchase_request_lines] api.one deprecated
purchase_request_to_rfq/models/purchase_order.py:116: [W8104(api-one-deprecated), PurchaseOrderLine.copy] api.one deprecated
************* Module purchase_request_to_rfq.models.purchase_request
purchase_request_to_rfq/models/purchase_request.py:40: [W8104(api-one-deprecated), PurchaseRequestLine._get_is_editable] api.one deprecated
purchase_request_to_rfq/models/purchase_request.py:46: [W8104(api-one-deprecated), PurchaseRequestLine._purchased_qty] api.one deprecated
purchase_request_to_rfq/models/purchase_request.py:55: [W8104(api-one-deprecated), PurchaseRequestLine._get_purchase_state] api.one deprecated
purchase_request_to_rfq/models/purchase_request.py:84: [W8104(api-one-deprecated), PurchaseRequestLine.copy] api.one deprecated
************* Module purchase_request_to_requisition.models.purchase_request
purchase_request_to_requisition/models/purchase_request.py:22: [W8104(api-one-deprecated), PurchaseRequestLine._get_is_editable] api.one deprecated
purchase_request_to_requisition/models/purchase_request.py:28: [W8104(api-one-deprecated), PurchaseRequestLine._requisition_qty] api.one deprecated
purchase_request_to_requisition/models/purchase_request.py:37: [W8104(api-one-deprecated), PurchaseRequestLine._get_requisition_state] api.one deprecated
purchase_request_to_requisition/models/purchase_request.py:72: [W8104(api-one-deprecated), PurchaseRequestLine.copy] api.one deprecated
************* Module purchase_request_to_requisition.models.purchase_requisition
purchase_request_to_requisition/models/purchase_requisition.py:81: [W8104(api-one-deprecated), PurchaseRequisitionLine._has_purchase_request_lines] api.one deprecated
purchase_request_to_requisition/models/purchase_requisition.py:98: [W8104(api-one-deprecated), PurchaseRequisitionLine.copy] api.one deprecated
************* Module purchase_request.models.purchase_request
purchase_request/models/purchase_request.py:127: [W8104(api-one-deprecated), PurchaseRequest.copy] api.one deprecated
purchase_request/models/purchase_request.py:144: [W8104(api-one-deprecated), PurchaseRequest.write] api.one deprecated
purchase_request/models/purchase_request.py:189: [W8104(api-one-deprecated), PurchaseRequestLine._get_supplier] api.one deprecated
```

Do you have planned migrate them to `api.multi`?

JonathanNEMRY commented:
@jbeficent could you please make a rebase of your branch? Thanks!

JordiBForgeFlow committed:
Merge branch '8.0-purchase_request' of https://github.com/Eficent/purchase-workflow into 8.0-purchase_request

Conflicts:
	purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py

fixes

JordiBForgeFlow commented:
@JonathanNEMRY done.

JonathanNEMRY commented:
Thanks @jbeficent 
And it is now green! So wait about some  :+1: to merge it :smile:

JordiBForgeFlow commented:
@kittiu You have been involve in this. Can you :+1: ?

kittiu commented:
:+1: 
It is important module in our project.

moylop260 commented on README.md:
```diff
@@ -11,3 +11,51 @@ This project aim to deal with modules related to manage purchase and their relat
  - Add validation step
  - ...
+
+[//]: # (addons)
```
This section is auto-computed... FYI you don't need this change here.

moylop260 commented:
> The following modules are added:
> purchase_request
> purchase_request_to_requisition
> purchase_request_to_rfq

![is_right](https://cloud.githubusercontent.com/assets/6644187/15969106/1812a3e0-2ef4-11e6-8bd4-84ab88b947dd.png)

JordiBForgeFlow commented:
@moylop260 Not sure why there are so many files in the PR, but when I do a 'git diff upstream/8.0' I only see changes made to the purchase request modules

sbidoul commented:
@moylop260 I did a checkout of @jbeficent 's branch and confirm it only adds the new modules without touching others. So I guess it's a github bug.

moylop260 commented:
@sbidoul 
Ok, I got it thank you
I saw the same case...

@jbeficent 
Thanks for contribution
Great work!

I saw some small checks... I have created the following list, but feel free of tell us if you wish fix them in this PR or not...

Please, could you consider the [following cases](https://travis-ci.org/OCA/purchase-workflow/jobs/136656861#L419-L440)?
[Odoo recommend us not use api.one](https://github.com/odoo/odoo/pull/8527)

``` bash
************* Module purchase_request_procurement.models.procurement
purchase_request_procurement/models/procurement.py:16: [W8104(api-one-deprecated), Procurement.copy] api.one deprecated
************* Module purchase_request_to_rfq.models.purchase_order
purchase_request_to_rfq/models/purchase_order.py:98: [W8104(api-one-deprecated), PurchaseOrderLine._has_purchase_request_lines] api.one deprecated
purchase_request_to_rfq/models/purchase_order.py:116: [W8104(api-one-deprecated), PurchaseOrderLine.copy] api.one deprecated
************* Module purchase_request_to_rfq.models.purchase_request
purchase_request_to_rfq/models/purchase_request.py:40: [W8104(api-one-deprecated), PurchaseRequestLine._get_is_editable] api.one deprecated
purchase_request_to_rfq/models/purchase_request.py:46: [W8104(api-one-deprecated), PurchaseRequestLine._purchased_qty] api.one deprecated
purchase_request_to_rfq/models/purchase_request.py:55: [W8104(api-one-deprecated), PurchaseRequestLine._get_purchase_state] api.one deprecated
purchase_request_to_rfq/models/purchase_request.py:84: [W8104(api-one-deprecated), PurchaseRequestLine.copy] api.one deprecated
************* Module purchase_request_to_requisition.models.purchase_request
purchase_request_to_requisition/models/purchase_request.py:22: [W8104(api-one-deprecated), PurchaseRequestLine._get_is_editable] api.one deprecated
purchase_request_to_requisition/models/purchase_request.py:28: [W8104(api-one-deprecated), PurchaseRequestLine._requisition_qty] api.one deprecated
purchase_request_to_requisition/models/purchase_request.py:37: [W8104(api-one-deprecated), PurchaseRequestLine._get_requisition_state] api.one deprecated
purchase_request_to_requisition/models/purchase_request.py:72: [W8104(api-one-deprecated), PurchaseRequestLine.copy] api.one deprecated
************* Module purchase_request_to_requisition.models.purchase_requisition
purchase_request_to_requisition/models/purchase_requisition.py:81: [W8104(api-one-deprecated), PurchaseRequisitionLine._has_purchase_request_lines] api.one deprecated
purchase_request_to_requisition/models/purchase_requisition.py:98: [W8104(api-one-deprecated), PurchaseRequisitionLine.copy] api.one deprecated
************* Module purchase_request.models.purchase_request
purchase_request/models/purchase_request.py:127: [W8104(api-one-deprecated), PurchaseRequest.copy] api.one deprecated
purchase_request/models/purchase_request.py:144: [W8104(api-one-deprecated), PurchaseRequest.write] api.one deprecated
purchase_request/models/purchase_request.py:189: [W8104(api-one-deprecated), PurchaseRequestLine._get_supplier] api.one deprecated
```
- Use [short license format](https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py#L1-L3)
- Execute [isort](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#imports)
- [If the technical name of the field is the same to the string of the label, don't put string parameter](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#field)
- For default date use `default=fields.Date.context_today`
- The compute methods should be called [_compute_field_name](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#field)
- The docstring avoid a newline at start and spaces and use `"""` [pep-257](https://www.python.org/dev/peps/pep-0257/#specification)
- Change `for request_id in requests_dict.keys():` by `for request_id in requests_dict:`
  similar case in purchase_request_to_rfq/models/purchase_order.py:68 and purchase_request_to_rfq/models/stock.py:71
- Change 

``` python
        if self.purchase_request_lines:
            self.has_purchase_request_lines = True
        else:
            self.has_purchase_request_lines = False
```

by `self.has_purchase_request_lines = bool(self.purchase_request_lines)`
Similar case in purchase_request_to_rfq/models/purchase_order.py:99
- purchase_request_to_requisition/wizard/purchase_request_line_make_purchase_requisition.py:48: [W0311(bad-indentation), ] Bad indentation. Found 16 spaces, expected 12
  purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:66: [W0311(bad-indentation), ] Bad indentation. Found 16 spaces, expected 12
  purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:213: [W0311(bad-indentation), ] Bad indentation. Found 24 spaces, expected 20
  purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:214: [W0311(bad-indentation), ] Bad indentation. Found 24 spaces, expected 20
  purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:215: [W0311(bad-indentation), ] Bad indentation. Found 24 spaces, expected 20
- Execute a `git rebase` in order to re-index the current github bug.

chrisdec commented:
+1 (functional tests)

JordiBForgeFlow commented:
Thanks @moylop260. We will review your points before merge. Also we have spotted a couple of issues testing on runbot that we need to fix before merging. @aheficent

JonathanNEMRY commented:
@jbeficent did you report those issues on github?
As we use this module into our projects then we're intrested  to know if there are problematic use cases

sbidoul commented:
@jbeficent are the last issues you have identified blocking? Or do they imply a change in data model that would require migration?

If not I'd suggest to merge as is, so we can e.g. start working on translations on transifex.

AaronHForgeFlow committed:
[FIX] pull189#issuecomment-225476547

JonathanNEMRY commented on purchase_request/models/purchase_request.py:
```diff
@@ -64,7 +49,7 @@ def _default_picking_type(self):
                  'line_ids', 'picking_type_id')
-    def _get_is_editable(self):
+    def _compute_is_editable(self):
         if self.state in ('to_approve', 'approved', 'rejected'):
```
@api.multi  should have a loop on self

JonathanNEMRY commented on purchase_request_to_requisition/models/purchase_request.py:
```diff
@@ -17,24 +17,27 @@
-    def _get_is_editable(self):
-        super(PurchaseRequestLine, self)._get_is_editable()
+    def _compute_is_editable(self):
+        self.ensure_one()
```
@api.multi should have a loop on self and remove self.ensure_one() it will give a stack trace if the field is shown into a tree view

JonathanNEMRY commented on purchase_request_to_requisition/models/purchase_request.py:
```diff
@@ -17,24 +17,27 @@
     @api.depends('requisition_lines.requisition_id.state')
-    def _get_requisition_state(self):
+    def _compute_requisition_state(self):
+        self.ensure_one()
```
same here

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_order.py:
```diff
@@ -94,33 +78,22 @@ def wkf_confirm_order(self):
-            self.has_purchase_request_lines = False
+    @api.multi
+    def _compute_has_purchase_request_lines(self):
+        self.ensure_one()
```
same here

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_request.py:
```diff
@@ -35,24 +21,27 @@ class PurchaseRequestLine(models.Model):
-    def _get_is_editable(self):
-        super(PurchaseRequestLine, self)._get_is_editable()
+    def _compute_is_editable(self):
+        self.ensure_one()
```
same comment here too

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_request.py:
```diff
@@ -35,24 +21,27 @@ class PurchaseRequestLine(models.Model):
-    @api.one
+    @api.multi
     def _purchased_qty(self):
+        self.ensure_one()
```
remove ensure_one and loop into the method

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_request.py:
```diff
@@ -35,24 +21,27 @@ class PurchaseRequestLine(models.Model):
     @api.depends('purchase_lines.state')
-    def _get_purchase_state(self):
+    def _compute_purchase_state(self):
+        self.ensure_one()
```
remove ensure_one and loop into the method

JonathanNEMRY commented on purchase_request/models/purchase_request.py:
```diff
@@ -91,8 +76,7 @@ def _get_is_editable(self):
                                   "request.",
-                             default=lambda *args:
-                             time.strftime('%Y-%m-%d %H:%M:%S'),
+                             default=fields.Date.context_today,
```
remove time from import to avoid pylint errors

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_order.py:
```diff
@@ -94,33 +78,22 @@ def wkf_confirm_order(self):
         'purchase_order_line_id',
         'purchase_request_line_id',
-        'Purchase Request Lines', readonly=True)
+        'Purchase Request Lines', readonly=True, copy=False)
```
this one makes the pep8 80-characters overflowed :wink:

AaronHForgeFlow committed:
[FIX] pull189 flake8 issues

AaronHForgeFlow committed:
[FIX] pull189 flake8 issues

JonathanNEMRY commented on purchase_request/models/purchase_request.py:
```diff
@@ -172,7 +171,7 @@ def _compute_is_editable(self):
-    @api.one
+    @api.multi
     def _compute_supplier_id(self):
         if self.product_id:
```
With @api.multi here it has to be something like

```
for rec in self:
    if rec.product_id:
```

JonathanNEMRY commented on purchase_request_to_requisition/models/purchase_request.py:
```diff
@@ -20,7 +20,6 @@ class PurchaseRequestLine(models.Model):
     @api.multi
     @api.depends('requisition_lines')
     def _compute_is_editable(self):
-        self.ensure_one()
```
Removing the ensure_one is not enough. You have to loop on self as my previous comment suggest it. This is required with the association of  @api.multi and the use of self

JonathanNEMRY commented on purchase_request_to_requisition/models/purchase_requisition.py:
```diff
@@ -77,12 +77,9 @@ def _prepare_purchase_order_line(self, requisition, requisition_line,
-            self.has_purchase_request_lines = True
-        else:
-            self.has_purchase_request_lines = False
+        self.has_purchase_request_lines = bool(self.purchase_request_lines)
```
same here your have to loop on self

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_order.py:
```diff
@@ -80,7 +80,6 @@ class PurchaseOrderLine(models.Model):
     @api.multi
     def _compute_has_purchase_request_lines(self):
-        self.ensure_one()
         self.has_purchase_request_lines = bool(self.purchase_request_lines)
```
same comment here

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_request.py:
```diff
@@ -24,14 +24,12 @@ class PurchaseRequestLine(models.Model):
     @api.multi
     @api.depends('purchase_lines')
     def _compute_is_editable(self):
-        self.ensure_one()
```
same comment here

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_request.py:
```diff
@@ -24,14 +24,12 @@ class PurchaseRequestLine(models.Model):
-    def _purchased_qty(self):
-        self.ensure_one()
+    def _compute_purchased_qty(self):
         purchased_qty = 0.0
```
same here

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_request.py:
```diff
@@ -41,7 +39,6 @@ def _purchased_qty(self):
     @api.multi
     @api.depends('purchase_lines.state')
     def _compute_purchase_state(self):
-        self.ensure_one()
```
same comment

AaronHForgeFlow committed:
[FIX] pull189 discussion_r68006610

JonathanNEMRY commented on purchase_request_to_requisition/models/purchase_request.py:
```diff
@@ -21,36 +21,38 @@ class PurchaseRequestLine(models.Model):
-                      for pr_line in self.requisition_lines]):
-                self.requisition_state = 'draft'
+        for rec in self:
+            rec.requisition_state = 'none'
```
knowing that `recordset.fields = 'value'` will  involve a `models.write()` I think it will be more clean to use a temporary variable that contains the `requisition_state` and then make the `self.requisition_state = temp` in the end. What do you think?

JonathanNEMRY commented on purchase_request_to_rfq/models/purchase_request.py:
```diff
@@ -25,34 +25,37 @@ class PurchaseRequestLine(models.Model):
-                      self.purchase_lines]):
-                self.purchase_state = 'draft'
+        for rec in self:
+            rec.purchase_state = 'none'
```
same comment than `_compute_requisition_state`

AaronHForgeFlow committed:
[FIX] pull189 discussion_r68008046

JonathanNEMRY commented:
With the last modifications (thanks @aheficent) :+1:  code review

AaronHForgeFlow commented:
Thanks  @JonathanNEMRY Very helpful ;)

JordiBForgeFlow commented:
@sbidoul This one is ready to merge.

moylop260 commented:
We have a opportunity area to add tests for those modules...
But for now we have many approvals and many fixes!

github show many files... but running
`git fetch -p OCA +refs/pull/*/head:refs/pull/OCA/*`
`git diff --name-only oca/8.0..pull/oca/189`

Result: 62 files

Then merging.
